### PR TITLE
Redis connector nil ptr check

### PR DESCRIPTION
**Pull request by DavidCorn**
Return NoOp default scope when stats is nil

### CODE REVIEW

File: metrics/metrics.go [hunk 1]
```diff
@@ -33,3 +53,51 @@ type Timer interface {
 	// Stop reports time elapsed since the timer start to the recorder.
 	Stop()
 }
+
+// CheckStats checks whether stats is nil. If it is nil then return default scope
+func CheckIfNilStats(stats Scope) Scope {
```
[the remark this answers is not in the document]
Comment (DavidCorn) [reply metrics/metrics.go]: 🆗

File: connectors/redis/redigo.go [hunk 2]
```diff
@@ -90,11 +90,10 @@ func (c *simpleRedis) Shutdown() error {
 // Do is a proxy method that calls Redigo's `Do` method and returns its output. It remembers
 // to close connections taken from the pool
 func (c *simpleRedis) do(commandName string, args ...interface{}) (interface{}, error) {
-	if c.stats != nil {
-		t := c.stats.SubScope("redis").SubScope("latency").Timer(commandName)
-		t.Start()
-		defer t.Stop()
-	}
+	t := metrics.CheckIfNilStats(c.stats).SubScope("redis").SubScope("latency").Timer(commandName)
```
Comment (tiffanycheng) [inline connectors/redis/redigo.go]: call checkifnilstats in NewRedigoClient

File: metrics/no_op_stats.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
```
Comment (tiffanycheng) [inline metrics/no_op_stats.go]: tests?

File: metrics/no_op_stats.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+// NoOpScope scopes nothing
+type NoOpScope struct {}
+
+// NoOpCounter Counts nothing
+type NoOpCounter struct {}
+
+// NoOpTimer times nothing
+type NoOpTimer struct {}
+
+func (s *NoOpScope) Counter(name string) Counter {
```
Comment (eculver) [inline metrics/no_op_stats.go]: lint: missing docstring.

File: metrics/no_op_stats.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+// CheckStats checks whether stats is nil. If it is nil then return default scope
+func CheckIfNilStats(stats Scope) Scope {
+	if stats == nil {
+		return &NoOpScope{}
+	}
+	return stats
+}
+
+// NoOpScope scopes nothing
+type NoOpScope struct {}
```
Comment (eculver) [inline metrics/no_op_stats.go]: nit: this shouldn't be camel-cased

File: metrics/no_op_stats.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+func (s *NoOpScope) SubScope(name string) Scope {
+	return &NoOpScope{}
+}
+
+func (s *NoOpScope) Timer(name string) Timer {
+	return &NoOpTimer{}
+}
+
+func (c *NoOpCounter) Inc(delta int64) {
+	// do nothing
```
Comment (eculver) [inline metrics/no_op_stats.go]: nit: move these to the docstring: `// Inc is a noop`

File: metrics/no_op_stats.go (hunk 3 above)
Comment (eculver) [inline metrics/no_op_stats.go]: suggest calling this just `noop.go` for simplicity.

File: metrics/no_op_stats.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import (
+	"time"
+)
+
```
Comment (eculver) [inline metrics/no_op_stats.go]: run `gofmt` -- extra lines will go away.

File: connectors/redis/redis.go [hunk 8]
```diff
@@ -96,7 +96,7 @@ func NewConnector(config Config, scope metrics.Scope) dosa.Connector {
 	return &Connector{
 		client: NewRedigoClient(config.ServerSettings, scope),
 		ttl:    config.TTL,
-		stats:  scope,
+		stats:  metrics.CheckIfNilStats(scope),
```
Comment (eculver) [inline connectors/redis/redis.go]: When would it make sense to pass a `nil` value here? Why not make`scope` required and if its value is `nil`, panic?

Comment (DavidCorn) [reply connectors/redis/redis.go]: We want the scope operation to work as normal even it is nil, by adding the CheckIfNilStats, we don't need to check status every time we want to use the `stats`.

File: metrics/no_op_stats.go (hunk 3 above)
Comment (DavidCorn) [reply metrics/no_op_stats.go]: Yeah will add test if the logic looks fine

**Pull request merged**